KAFKA-10049: Fixed FKJ bug where wrapped serdes are set incorrectly when using default StreamsConfig serdes

bellemare opened the pull request:
Bug Details:
Mistakenly setting the value serde to the key serde for an internal wrapped serde in the FKJ workflow. 


Testing:
Added integration test to use a non-primitive Serde, in this case the JSONSerde that the original bug finder reported using. Expanded integration test to ensure that the default Serdes work for the entire happy path of the FKJ.

Introduces a testing dependency on com.fasterxml.jackson, though this is already the case in other modules so I suspect it won't be a big issue.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinMaterializationIntegrationTest.java:
```diff
@@ -77,11 +83,12 @@ public KTableKTableForeignKeyJoinMaterializationIntegrationTest(final boolean ma
-            mkEntry(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath())
-        ));
+        streamsConfig = new Properties();
+        streamsConfig.put(StreamsConfig.APPLICATION_ID_CONFIG, "my-stream-processing-application-2");
```
Please use the safeTestName

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinMaterializationIntegrationTest.java:
```diff
@@ -94,15 +101,49 @@ public void before() {
     }
 
+    @Test
+    public void shouldEmitRecordWhenJoiningWithDefaultSerdes() {
```
Hey @bellemare , do we need to add a new test? It looks like this condition was intended to be covered already in KTableKTableForeignKeyJoinScenarioTest . Should we just fix that test instead?

bellemare replied:
Okay, will check.

bellemare replied:
@vvcephei Just getting back to this now. Before I rework KTableKTableForeignKeyJoinScenarioTest:

I can keep it simple and try something like:
KTable<String, Long> aTable
KTable<Long, Integer> bTable

Or I can go ahead and use the JSONSerde, User, and Foo classes. I think the former will suffice for coverage purposes, but I like the latter because it may better reflect real-life use-cases (and I usually look to the tests myself to show me how to use a function). Thoughts on this before I refactor?

bellemare replied:
Went with "keep it simple" basic Integer and String.

vvcephei reviewed (changes requested):
Thanks, @bellemare , just a high-level question about the testing strategy.

guozhangwang reviewed (commented):
@bellemare could you point to me which line is the actual fix?

guozhangwang commented:
test this please

bellemare commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:
```diff
@@ -66,10 +66,13 @@ public void init(final InternalProcessorContext context) {
         // then pass it the default one from the context
         if (valSerializer instanceof WrappingNullableSerializer) {
-            ((WrappingNullableSerializer) valSerializer).setIfUnset(context.valueSerde().serializer());
+            ((WrappingNullableSerializer) valSerializer).setIfUnset(
```
@guozhangwang This is 1/2 of the areas that needed the fix. The valueSerde was being passed into the underlying Serde, despite it really needing the keySerde.

bellemare commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java:
```diff
@@ -87,10 +87,13 @@ public void init(final InternalProcessorContext context) {
         // then pass it the default one from the context
         if (valDeserializer instanceof WrappingNullableDeserializer) {
-            ((WrappingNullableDeserializer) valDeserializer).setIfUnset(context.valueSerde().deserializer());
+            ((WrappingNullableDeserializer) valDeserializer).setIfUnset(
```
@guozhangwang This is 2/2 of the areas that needed the fix. Same thing, the valueSerde was being passed into the underlying Serde, despite it really needing the keySerde.

guozhangwang replied:
Ack, thanks!

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java:
```diff
@@ -63,9 +63,9 @@ public SubscriptionWrapperSerde(final Supplier<String> primaryKeySerializationPs
+        public void setIfUnset(final Serializer<K> defaultKeySerializer, final Serializer<Void> defaultValueSerializer) {
             if (primaryKeySerializer == null) {
-                primaryKeySerializer = Objects.requireNonNull(defaultSerializer, "defaultSerializer cannot be null");
+                primaryKeySerializer = Objects.requireNonNull(defaultKeySerializer);
```
@guozhangwang , this (and L128) is the "actual fix". Previously, the actual default serde we would pass in was the value serde (in SinkNode/SourceNode), but this one particular serde needs the key serde. If you look at the other serdes that changed in this diff, you'll note that they all actually do need the value serde. Only this one is different.

vvcephei replied:
Oh, hah. Looks like you beat me to it, @bellemare  :)

bellemare commented:
Updated KTableKTableForeignKeyJoinScenarioTest from (String,String) to (Integer,String). This would have caught this particular bug. Reverted the JSON-related test since it's not essential for our coverage.

vvcephei commented:
Test this please

vvcephei commented:
Thanks, @bellemare ! I'll give it a look today.

vvcephei commented:
Looks like checkstyle failed. I may submit a patch, just to get the tests running.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:
```diff
@@ -243,17 +244,17 @@ private void validateTopologyCanProcessData(final StreamsBuilder builder) {
+            final TestInputTopic<Integer, String> aTopic = topologyTestDriver.createInputTopic("A", new IntegerSerializer(), new StringSerializer());
+            final TestInputTopic<Integer, String> bTopic = topologyTestDriver.createInputTopic("B", new IntegerSerializer(), new StringSerializer());
+            final TestOutputTopic<Integer, String> output = topologyTestDriver.createOutputTopic("output", new IntegerDeserializer(), new StringDeserializer());
+            aTopic.pipeInput(1, "1-alpha");
+            bTopic.pipeInput(1, "beta");
```
Might be nice for demonstration purposes if the two records actually have different keys. Maybe: 

```suggestion
            aTopic.pipeInput(1, "999-alpha");
            bTopic.pipeInput(999, "beta");
```

bellemare replied:
Agreed. Good catch.

vvcephei replied:
bump (it looks like this didn't get changed)

bellemare replied:
I swear I know how to read... T_T

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:
```diff
@@ -181,60 +182,60 @@ public void shouldWorkWithDefaultAndProducedSerdes() {
+                applicationId + "-right_table-STATE-STORE-0000000003-changelog--value",
+                // output topics
+                "output-topic--key",
+                "output-topic--value"
```
It'd be better to avoid reformatting code to change indentation settings. AFAIK, in Streams at least, we tend to use indents of 4 spaces.

bellemare replied:
Right, sorry, my bad.

vvcephei reviewed (commented):
Thanks for the update, @bellemare ! Just a couple of minor comments on the tests, but it looks great overall.

vvcephei commented:
Ok, since I had a couple of small requests, I won't push a fix for the checkstyle errors. In case they're gone by the time you come back to this, it was:
```
00:11:48.980 > Task :streams:checkstyleTest
00:11:48.981 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:71: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:48.981 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:72: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:48.981 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:78: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:48.982 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:94: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:48.982 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:95: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.132 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:101: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.132 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:117: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.132 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:118: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:126: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:142: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:143: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:149: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:166: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:167: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.133 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:173: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.134 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:206: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
00:11:49.134 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:207: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
```

By the way, if you run `./gradlew :streams:test`, it includes the checkstyle rules. It takes a while to run the tests, but it's faster than waiting for Jenkins. Often, if I don't want to wait before pushing to the branch, I'll run it on my machine after pushing, just so I'll catch stuff like this.

bellemare committed:
KAFKA-10049: Fixed FKJ bug where wrapped serdes are set incorrectly when using default StreamsConfig serdes

bellemare commented:
I swear it ran checkstyle when I compiled it... bah! I guess not. I did run ./gradlew :streams:test it had one unrelated error (I think) so lets see how it does now.

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
retest this please

bellemare committed:
KAFKA-10049: Changed primary key from 1 to 999 to further clarify the join mechanisms

bellemare commented:
retest this please

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Hey @bellemare , Thanks so much for the update! Jenkins has been a bit lazy recently, so I'll probably be pinging it for a while...

vvcephei commented:
Ah, there we go :)

vvcephei commented:
I've noticed that if you complain, you tend to get results...

vvcephei commented:
All the test failures were `kafka.admin.ReassignPartitionsUnitTest.testModifyBrokerThrottles`

bellemare commented:
Thanks John!

vvcephei commented:
Np, Thanks for the fix, @bellemare !

vvcephei commented:
Cherry-picked to 2.6 and 2.5
